File: Warmup/vw07p.cpp
#include <iostream>
#include <vector>
using namespace std;

const int high = 1e7;

int n, m;
vector <int> tm, vtm;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie();
    cout.tie();
    
    cin >> n >> m;
    tm.resize(n);
    vtm.resize(m + 1, high);

    cin >> tm[0];
    for (int i = 1; i < n; i++) 
    {
        cin >> tm[i];
        tm[i] += tm[i - 1];
    }

    vtm[m] = 0;
    for(int i = n; i >= 1; i--)
    {
		for(int j = m; j >= 0; j--)
        {
			if (vtm[j] == 1e9 || j - tm[i - 1] < 0) continue;
			vtm[j - tm[i - 1]] = min(vtm[j - tm[i - 1]], vtm[j] + i + 1);
		}
	}

    int overtime = tm[n - 1] - tm[n - 2];

    for(int j = m; j >= 0; j--)
    {
		if(vtm[j] == 1e9) continue; 
		if(j - tm[n - 1] >= 0 && (j - tm[n - 1]) % overtime == 0)
			vtm[0] = min(vtm[0], vtm[j] + n + 1 + (j - tm[n - 1]) / overtime);
	}

    if (vtm[0] == high) cout << "-1"; else cout << vtm[0] - 1;

    return 0;
}

File: Warmup/PythonCode/vw07p.py
n, m = map(int, input().split())
a = list(map(int, input().split()))

ax = []
ax.append(a[0])
for i in range(1, n):
    ax.append(ax[i - 1] + a[i])

vtm = [int(1e7) for i in range(m)]
vtm.append(0)
for i in range(n, 0, -1):
    for j in range(m, -1, -1):
        if vtm[j] == int(1e7) or j - ax[i - 1] < 0: continue
        vtm[j - ax[i - 1]] = min(vtm[j - ax[i - 1]], vtm[j] + i + 1)

overtime = ax[n - 1] - ax[n - 2]
for i in range(m, -1, -1):
    if vtm[i] == int(1e7): continue
    if j - ax[n - 1] >= 0 and (j - ax[n - 1]) % overtime == 0:
        vtm[0] = min(vtm[0], vtm[j] + n + 1 + (j - ax[n - 1]) / overtime)

if vtm[0] == int(1e7): print("-1")
else: print(vtm[0] - 1)